onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MIN:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMin<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+    case AGGREGATE_FUNCTION::MAX:
+      compute_agg(
+          X, Z, label,
+          _AggregatorMax<ITYPE, OTYPE>(
+              roots_.size(), n_targets_or_classes_,
+              post_transform_, &(base_values_)));
+      return;
+  }
+  ORT_THROW("Unknown aggregation function in TreeEnsemble.");
+}
+
+template <typename ITYPE, typename OTYPE>
+template <typename AGG>
+void TreeEnsembleCommon<ITYPE, OTYPE>::compute_agg(const Tensor* X, Tensor* Z, Tensor* label, const AGG& agg) const {
+  int64_t stride = X->Shape().NumDimensions() == 1 ? X->Shape()[0] : X->Shape()[1];
+  int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
+
+  const ITYPE* x_data = X->template Data<ITYPE>();
+  OTYPE* z_data = Z->template MutableData<OTYPE>();
+  int64_t* label_data = label == NULL ? NULL : label->template MutableData<int64_t>();
+
+  if (n_targets_or_classes_ == 1) {
+    if (N == 1) {
+      OTYPE scores = 0;
+      unsigned char has_scores = 0;
+      if (n_trees_ <= omp_tree_) {
+        for (int64_t j = 0; j < n_trees_; ++j)
+          agg.ProcessTreeNodePrediction1(
+              &scores,
+              ProcessTreeNodeLeave(roots_[j], x_data),
+              &has_scores);
+      } else {
+        std::vector<OTYPE> scores_t(n_trees_, 0);
+        std::vector<unsigned char> has_scores_t(n_trees_, 0);
+#ifdef USE_OPENMP
+#pragma omp parallel for
+#endif
+        for (int64_t j = 0; j < n_trees_; ++j) {
+          agg.ProcessTreeNodePrediction1(
+              &(scores_t[j]),
+              ProcessTreeNodeLeave(roots_[j], x_data),
+              &(has_scores_t[j]));
+        }
+        auto it = scores_t.cbegin();
+        auto it2 = has_scores_t.cbegin();
+        for (; it != scores_t.cend(); ++it, ++it2)
+          agg.MergePrediction1(&scores, &has_scores, &(*it), &(*it2));
+      }
+
+      agg.FinalizeScores1(z_data, scores, has_scores, label_data);
+    } else {
+      if (N <= omp_N_) {
+        OTYPE scores;
+        unsigned char has_scores;
+        size_t j;
+
+        for (int64_t i = 0; i < N; ++i) {
+          scores = 0;
+          has_scores = 0;
+          for (j = 0; j < static_cast<size_t>(n_trees_); ++j)
+            agg.ProcessTreeNodePrediction1(
+                &scores,
+                ProcessTreeNodeLeave(roots_[j], x_data + i * stride),
+                &has_scores);
+          agg.FinalizeScores1(z_data + i * n_targets_or_classes_, scores, has_scores,
+                              label_data == NULL ? NULL : (label_data + i));
+        }
+      } else {
+        OTYPE scores;
+        unsigned char has_scores;
+        size_t j;
+
```

Comment:
What's the reason for declaring these outside the loop?

Reply:
To avoid the creation of these variables at every iteration. If it is parallelized, you will get a variable for every thread.

Reply:
Out of interest can you help me understand what the effective difference is between an assignment inside every loop vs. constructing inside the loop? Naively I would expect you're still writing a value to a location or locations of a certain size, whether that's via assignment or construction.

Reply:
I don't know how the compiler is handling this and looking at the assembly would be best to see if there is a difference. However, the basic scenario is when you do a parallelized sum, you need a variable sum in each thread to accumulate and a final one to receive the sum of all sum per thread. Having only one sum means using synchronisation at every iteration and not only at the end. scores and has_cores are not simple sums but are used in the same way (accumulators).